core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java

```diff
@@ -96,7 +96,10 @@ private static PartitionMap<PartitionStats> collectStats(
         StructLike key = keyTemplate.copyFor(coercedPartition);
         Snapshot snapshot = table.snapshot(entry.snapshotId());
         PartitionStats stats =
-            statsMap.computeIfAbsent(specId, key, () -> new PartitionStats(key, specId));
+            statsMap.computeIfAbsent(
+                specId,
+                ((PartitionData) file.partition()).copy(),
```

Comment:
I'm a bit concerned about these casts.

Could `keyTemplate.copyFor(file.partition())` work?

Reply:
hi @pvary, thanks for taking a look.
`keyTemplate` is of unified tuple struct, so it can't be used. btw, if you grep iceberg code, you'll find a number of same casts.

Reply:
Seems like the tests are broken by the infra 😢 

Could you help me understand the original issue? I thought that I understand, but seems like I still miss something.

I get that there is difference between the `partitionType` (which is the unified spec), the `coercedPartition` which is the partition key for the unified partition type.

What is the key we would like to see in the `statsMap`?

Reply:
discussed in private. @pvary, please let me know if any questions left.

Reply:
Discussed this with @RussellSpitzer, and he confirmed that casting is indeed correct.